Feature 11/add alchemy folder

Jinlee057 opened the pull request:
1. owners_for_collection
   - opensea에서 8/9일 조회시 Top NFT> All 에 있던 두 컬렉션 BAYC와 Pudgy Penguins을 소유하고 있는 지갑주소를 받아옵니다.
   - 지갑주소, 컨트랙트주소, 생성일 3가지 정보만 DB에 적재합니다 테이블명: alchemy_owners_for_contract
2. collections_for_owner
   - 추후에는 실제 고래 지갑주소로 대체 될 예정이나, 모델링 가능 여부만 판단하기 위해서 1번 이용 적재된 지갑주소를 활용합니다.
   - 실제 1번에서 나온 유니크한 지갑주소는 9900개 정도였으나, 테스트용으로 200개만 선정. (지갑에 대한 정보가 아직 없어 적재된 순서로 200개의 지갑)
   - 지갑주소, 소유 컬렉션명, opensea slug, opensea floor price, description ...(DB 참조)적재합니다 테이블명 : alchemy_collection_for_owner
  3. 저장되어 있는 코드로 깃헙액션 등록 및 실행됨을 확인했습니다.
  4. VM실행 여부는 추후 진행예정입니다.

Jinlee057 committed:
feat: Add get_monthly_top100 #11

Jinlee057 committed:
feat:Upload alchemy file and folder

Jinlee057 committed:
feat: Upload alchemy file and folder #11

upload temp task code. need to be amended with flow

Jinlee057 committed:
feat: Update alchemy flow #11

Upload flow file for deployment and test
Remove get_montly_top100.py from mart folder
Add get_monthly_top100, get_db_data task into utils
Add alchemy tasks

Jinlee057 committed:
feat: Update utils.py remove 2 tasks #11

Jinlee057 committed:
feat: Update alchemy_daily.py add 2tasks #11

Jinlee057 committed:
feat: Update alchemy_daily.py #11

Jinlee057 committed:
feat: Update alchemy_extract_task.py #11

Jinlee057 committed:
feat: Update init files #11

DataHyeon commented on .github/workflows/workflow_dispatch_flow_alchemy.yml:
워크 플로우 파일은 굳이 새로 안 만드셔도 됩니다
기존에 있던 워크플로우와 다른 점이 있다면 알려주세요

Jinlee057 replied:
없습니다 실습과정에서 실수로 만든 파일이고 삭제해야합니다

DataHyeon commented on flow/warehouse/alchemy_daily.py:
```diff
@@ -0,0 +1,61 @@
+def create_db_owners_for_contract():
+    
+    engine = get_sql_engine('gcp-mlops-sql-postgres').get_engine()
+    contracts=['0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D','0xBd3531dA5CF5857e7CfAA92426877b022e612cf8']
```
test_alchemy.py에서도 동일하게
이렇게 컨트랙트 주소 안에다가 하드 코딩하지 마시고
바깥에서 매개변수로 받도록 하는게 좋을 거 같네요

Jinlee057 replied:
아하 넵 수정반영하겠습니다

DataHyeon commented on flow/warehouse/alchemy_daily.py:
지갑을 테스트로 뽑아와서 추천시스템에 쓸 데이터를 만드는 거라면 지금 풀리퀘스트를 하는 게 아니라
전체적으로 다 완성이 된 코드들을 가져와야겠죠??

정리하자면 샘플로 쓰려고 하는 데이터들을 굳이 실제 배포를 해서 할 필요는 없을 거 같습니다.

파이프 라인으로 만들고 싶다면 다른 컨트랙트 주소도 받아서 사용하게끔 하던지 아니면
앞단에서 인기가 있는 상위 컨트랙트들이 정해져 있는 상태에서 그걸 입력값으로 받아와 해당 작업들이 진행 되도록 해야할 거 같습니다.

Jinlee057 replied:
넵 확인했습니다

DataHyeon commented on module/warehouse/alchemy/task/extract_task.py:
```diff
@@ -0,0 +1,226 @@
+    return owners
+
+
+@task(log_prints=True)
```
이 부분부터 어떤 조건을 줘서 가져오는 거 같은데
이거는 따로 모듈을 만들던가 아니면 이전에 말씀드린 것처럼 이렇게 추출된 데이터들이 특별한 용도로 사용된다면 데이터 마트를 구성하도록 하는 파이프라인을 만들면 좋을 거 같습니다.

Jinlee057 replied:
쿼리문으로 db에서 데이터 가져올 때 쓰려고 만들었는데 다시 보니 내부에서 onwer 데이터로 지정하면 범용으로 쓸 수가 없는 코드네요 지금 현재 목적은 추천시스템 테스트를 위해서 지갑주소 200개 가져오는 역할만 하고 있고
모듈로 할지 데이터 마트로 할지 앞으로의 방향 고민 해보겠습니다

DataHyeon commented on module/warehouse/alchemy/task/extract_task.py:
```diff
@@ -0,0 +1,226 @@
+    return df_jun_top100
+
+
+#@task
```
사용 안하는 주석처리 된 레거시 코드는 지우세요~

Jinlee057 replied:
넵 지우겠습니다

DataHyeon commented on module/warehouse/alchemy/task/load_task.py:
try, except, else  구문을 왜 저렇게 사용하는지 아시나요?

Jinlee057 replied:
따로 찾아보지는 않았고 코드만 봤을 때 try해서 에러나면 except, 아니면 else 실행 하는 것으로 이해하고 넘어갔습니다.
load_to_db를 아직 코드내에서는 사용안했습니다
아직 헷갈리는게 많아서 적재하는 코드도 모듈화하지 않고 직접 입력해서 사용하고 있었습니다
추후 리팩톨이 통해서 모듈화 할 수 있는 부분들 진행하겠습니다

DataHyeon reviewed (changes requested):
전반적으로 구성을 잘 되어있습니다.
각 코드별 리뷰 확인하시고 수정하신 이후 다시 검토 해봅시다

Jinlee057 committed:
feat: Update alchey_daily.py #11

based on actual data, 2 more flows added.
load_get_top_seller : get top sellers using query and pass seller list to create_db_collecition_for_seller function
load_get_top_buyer : same way but for buyer

Jinlee057 committed:
feat: Update extract_task.py #11

delete function create_owner_for_collection as no longer used under current flow.
add function get_top_seller, get_top_buyer details added in comment.

Jinlee057 committed:
feat: Update transfor_task.py #11

update preprocess function to rename column by convention.
drop unnecessary column 'bannerImageUrl'.
convert data type for numeric columns.
add timestamp

Jinlee057 committed:
feat:Update extract_task.py #11

data pipeline flow amended. get top100 sellers and buyesr as a list from db table daily_whale_100.

Jinlee057 committed:
feat: Update alchemy_daily flow #11

get nft colleciton infomation using alchemy API. and load to DB.

Jinlee057 committed:
feat: Update extract_task.py #11

amend target_date to previous date(today-timedleta(days=1))

Jinlee057 committed:
feat: Update transfform_task.py #11

amend data_created_at. same date with duen query date instead of today

Jinlee057 committed:
feat: Update __init__.py #11

add tasks from extract, transform file

Jinlee057 committed:
feat: Delete test files #11

delete test_alchemy.py
delete landing_alchemy_daily.py which is not required anymore

Jinlee057 committed:
feat: Update transform_task.py #11

add if statement for col 'bannerImageUrl' to prevent crash

Jinlee057 committed:
feat: Update alchemy.py #11

alchemy_daily.py flow and tanksform_task.py task

Jinlee057 reviewed (commented):
1. 목데이터가 아닌 실제 데일리 top 100 seller, buyer 지갑 내에 있는 컬렉션 정보 추출 후
DB에 적재하도록 내용 수정하였습니다.
2. github action & docker image & vm 등록 통한 자동 배포 완료하였습니다.

검토 후 승인 부탁드립니다.

kangsumin2ya commented on module/warehouse/alchemy/task/extract_task.py:
```diff
@@ -0,0 +1,178 @@
+
+
+@task(log_prints=True)
+def make_request_with_backoff(url, max_retries=10, max_backoff=64):
```
이 부분이 alchemy에서 api 요청 횟수 조절하는 코드인가요?
  max_retries는 어떤 계산으로 나왔나요??

Jinlee057 replied:
네 alchemy에 동시에 api 요청 많아질 경우 발생하는 에러를 해결하기 위한 코드인데
현재 기준 api 통신 문제 발생하지 않아 사용하지 않고 있습니다.
max_tries는 직접 설정하는 부분으로 10번 정도 재시도 하면 되지 않을까 싶어 넣었습니다.

kangsumin2ya commented on module/warehouse/alchemy/task/extract_task.py:
```diff
@@ -0,0 +1,178 @@
+            if page_key == None:
+                data = {'owner_address':owners_list,'contract':[contract]*len(owners_list)}
+                df_owner=pd.DataFrame(data)
+                df_owners_for_contract = pd.concat([df_owners_for_contract,df_owner])
```
아래의 create_collection_for_owner함수의 ThreadPoolExecutor에선 extend를 사용하셨는데 여기에선 concat을 사용하신 이유가 궁금합니다.

kangsumin2ya reviewed (commented):
확인했습니다!!

DataHyeon commented on module/warehouse/alchemy/task/load_task.py:
```diff
@@ -0,0 +1,14 @@
+@task(log_prints=True)
+def load_to_db(engine:sqlalchemy.engine.base.Engine,df:pd.DataFrame,table_name:str):
+    try:
+        df.to_sql(table_name, engine, if_exists='append', index=False)
```
이후에는 sqlalchemy 메서드를 이용해서 테이블에 추가하는 방법도 구현해보면 좋을 거 같습니다

그리고 해당 테이블의 제약조건등을 추가해서 중복되는 데이터들을 필터링 하는 것도 해보시면 좋을 거 같습니다^^

DataHyeon commented on module/warehouse/alchemy/task/transform_task.py:
```diff
@@ -0,0 +1,24 @@
+    if 'bannerImageUrl' in df.columns:
+        df.drop(columns=['bannerImageUrl'], inplace=True)
+    
+    df['total_balance']=pd.to_numeric(df['total_balance'], errors='coerce')
```
여기서 데이터 타입을 numeric으로 변경하는 이유가
디비 데이터 타입을 초과해서 그런 거 같은데 나중에 테이블 생성할 때 내가 넣을 데이터 크기를 고려해서 미리 설정해놓으면 좋을 거 같습니다

너무 큰 데이터일 거 같다면 어제 멘토님 말씀처럼 처음부터 문자열 타입으로 지정해줘도 좋을 거 같구요

지금도 잘 하셨습니다!

DataHyeon reviewed (approved):
전반적으로 잘 하셨습니다
제가 작성한 피드백 보시구 다음 작업간에는 반영해서 시도해보셔요
